moc_openstack_api_reporting.py

```diff
@@ -0,0 +1,572 @@
+        user_dict[user.id] = user
+
+    project_dict = {}
+    for project in openstack_conn.list_projects():
+        project_dict[project.id] = project
+
+    volume_dict = {}
+    for volume in openstack_cinder.volumes.list(search_opts={"all_tenants": True}, detailed=True):
+        volume_dict[volume.id] = volume
+
+    events = []
+    server_dict = {}
+    ap_t = datetime.datetime.strptime(script_timestamp, date_format_str)
+    for server in openstack_nova.servers.list(search_opts={"all_tenants": True}, detailed=True):
+        server_dict[server.id] = compile_server_state(server, project_dict, flavor_dict, user)
+        event_data = {"event_type": "compute.instance.exists", "event_time": script_timestamp}
+        if server_t < ap_t:
+            ap_t = server_t
+        if server.id in vm_timestamps:
+            event_data["start_ts"] = vm_timestamps[server.id]
+
+        if server.status == "ACTIVE" or server.status == "DELETED":
+            event_data["audit_period_start"] = server.created
+            event_data["audit_period_end"] = script_timestamp
+            event_data["start_ts"] = server.created
+            server_t = datetime.datetime.strptime(server.created, "%Y-%m-%dT%H:%M:%SZ")
+            events.append(build_event(server_dict[server.id], event_data))
+            if server.status == "deleted":
+                del vm_timestamps[server.id]
+
+    for server_id, server in server_dict.items():
+        # Only get the data since we last ran the script
+        # event_list = openstack_nova.instance_action.list(server_id, changes_since=last_timestamp, changes_before=script_timestamp)
+        # event_list = openstack_nova.instance_action.list(server_id, changes_since=None, changes_before=None)
+        event_list = openstack_nova.instance_action.list(server_id)
+        for event in event_list:
+            event_data = {
+                "audit_period_start": ap_t.strftime(date_format_str),
+                "audit_period_end": script_timestamp,
+                "event_type": f"compute.instance.{event.action}",
+                "event_time": event.start_time,
+                "request_id": event.request_id,  # not certain if this is meaningful or that this is the request id xdmod is expecting
+            }
+            events += convt_to_ceilometer_event_types(build_event(server, event_data))
+            if server_id not in vm_timestamps:
+                vm_timestamps[server_id] = {}
+            vm_timestamps[server_id]["timestamp"] = event_data["event_time"]
+            vm_timestamps[server_id]["updated"] = 1
+
+    json_out = f"{config['outdir']}/{last_run_timestamp}_{script_timestamp}.json"
+    with open(json_out, "w+", encoding="utf-8") as outfile:
+        json.dump(events, outfile, indent=2, sort_keys=True, separators=(",", ": "))
+    print(f"output file: {json_out}")
+
+    for key, val in vm_timestamps.items():
+        if val["updated"] == 0:
+            del vm_timestamps[key]
```

Comment:
doesn't this produce a runtime error in python3 when a dictionary is modified while you are iterating over it?

Reply:
```
>>> vm_timestamps = {"A": 1, "B": 2, "C": 3, "D": 4}
>>> for key, val in vm_timestamps.items():
...     if val==2:
...             del vm_timestamps[key]
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: dictionary changed size during iteration
```

Why can't I do this then?

Reply:
I was not getting that runtime error.  will change this.

Reply:
> I was not getting that runtime error

maybe it didn't hit the branch where the delete takes place.

Reply:
I would investigate it further, but it is a waste of time, easier to pull the array of keys, then iterate though them to delete the ones that are no-longer needed.

I have already re-written it this way last friday, and haven't gotten around to testing it fully.

Reply:
> I have already re-written it this way last friday, and haven't gotten around to testing it fully.

will you commit the fix in this PR? I think if we iterate over `list(vm_timestamps.items())` then that should take care of things.